### DISCUSSION TITLE
Replacing header formatting

### DISCUSSION THREAD

**Discussion by sojusnik** [peter88213/zim2obsidian#2] [Ideas]
First of all many thanks for such a useful script!

When following your instructions to convert Zim notes to markdown through Zim's built-in export function, then header 1  is converted to

```
Header 1
=============
```

and header 2 to

```
Header 2
-------------
```

while header 3 to 5 are created with hashtags, f.i.

```
### Header 3
```

Because Obsidian also shows you the markdown code when editing text, headers with only hashtags look much cleaner than having lines, as in header 1 and 2.

Is it maybe possible, that your awesome script could also convert header 1 and 2 having hashtags instead of lines in the converted markdown files?

FYI, those are setex and atx headers, as described [here](https://daringfireball.net/projects/markdown/syntax#header).

**Comment by peter88213** [comment]
It's done. Since the pages are now processed line by line anyway, zim2obsitian is also converting the horizontal rulers with version 0.3.

**Comment by sojusnik** [comment]
Many thanks! Will convert soon and give some feedback.

**Comment by peter88213** [comment]
I'm glad you like it. However, after having a look at Obsidian, I decided to stick with Zim for now. 
So I'm afraid I can't put too much effort into developing the zim2obsidian script. But in the large Obsidian community there will surely be someone out there with Python knowledge if more features are desired. That's what open source software is all about, isn't it?

**Comment by sojusnik** [comment]
> However, after having a look at Obsidian, I decided to stick with Zim for now.

What's your reasons to return to Zim?

I'm a 10+ years heavy Zim user and find that Zim generally covers all the basic functions of an offline personal knowledge management system, but it still lacks contributions to keep up with the development of Obsidian, which is already more feature-rich than Zim, despite its young age. Although it hurts sentimentally to migrate, but in the long run Obsidian has just more potential, especially with its huge community and rich ecosystem of plugins, which gave birth to unique plugins like [Excalidraw](https://github.com/zsviczian/obsidian-excalidraw-plugin/), that reduces the gap between text and drawings (even the built-in [Canvas](https://obsidian.md/canvas) is a good start in this direction). And the mobile experience with the dedicated [Obsidian app](https://play.google.com/store/apps/details?id=md.obsidian) (which BTW should soon get a [quick capture mode](https://obsidian.md/roadmap/)) is far superior to Zim + [Markor](https://github.com/gsantner/markor/), which is often a major point in our world of interoperability.

Of course it all depends on individual preferences, but IMHO for the general user Obsidian should do it in the long run.

**Comment by peter88213** [comment]
> What's your reasons to return to Zim?

Zim (I'm still using version 0.69.1) just does the job for me, that's all. 

Obsidian, with all its possibilities to play and tinker with, is indeed tempting, but at the moment I just can't afford the time to explore its features, let alone get into plugin programming.

As for the Obsidian community, I have found the request for a Zim converter in various places, however no one else seems to have picked up on it in all these years. Strange, isn't it? Anyway, Obsidian is cool, so have a lot of fun with it!

**Comment by sojusnik** [comment]
> Obsidian, with all its possibilities to play and tinker with, is indeed tempting, but at the moment I just can't afford the time to explore its features, let alone get into plugin programming.

Was in the same situation, but decided to finally put some time aside to make the jump. The only thing that stops me from using Obsidian fully is the migration from Zim. At the moment, there seems to be no easy and automated solution. Your script is one of the most promising ones. Another one that I've found is [this](https://github.com/pschwede/zim2md) one. But it also has some issues. Let's see if the developer is interested in fixing them.

Let me know if you are interested in improving this script, because I've already found several issues. Maybe combining your script and the one from above (also written in python) would be the solution, because the other script is converting things properly, which are missing in yours and vice versa.

**Comment by peter88213** [comment]
Well, it looks like pschwede's script is to be applied directly to the zim notebook, whereas my script relies on the zim Markdown exporter to convert everything to a commonly understood Markdown format. 

Note that I didn't touch the markup created by the zim exporter at first (except the link style). But with re-formatting the headings, a line-by-line processing routine is introduced that can be extended for further conversions. Once you understand the method, this could be a beginner-level Python exercise. Maybe you'd like to pick up the spirit of the open source movement, and get involved with the code yourself?

This is the pattern in zim2obsidian.py:

```
        # Replace Setext-style headers with Atx-style headers; convert rulers.
        newLines = []
        previousLine = None
        for  line in lines:
            if line.startswith('=') and line.count('=') == len(line):
                print(f'Converting 1st level heading "{previousLine}" ...')
                previousLine = f'# {previousLine}'
            elif line.startswith('-') and line.count('-') == len(line):
                print(f'Converting 2nd level heading "{previousLine}" ...')
                previousLine = f'## {previousLine}'
            elif line.startswith('*') and line.count('*') == len(line):
                print('Converting horizontal ruler ...')
                if previousLine is not None:
                    newLines.append(previousLine)
                previousLine = '---'

            # Just add another conversion here (shown as pseudo-code):

            elif (line contains other markup to convert):
                if previousLine is not None:
                    newLines.append(previousLine)

                previousLine = (line, old markup replaced with new markup) 
            
            #######

            else:
                if previousLine is not None:
                    newLines.append(previousLine)
                previousLine = line
```

**Comment by peter88213** [comment]
> Let me know if you are interested in improving this script, because I've already found several issues. 

Well, all I can do is generously comment my code so that someone else will have an easier time understanding and extending it. 
This has been done with version 0.3.3. 

With version 0.4.0, the script is organized as a Python module that can be imported into the test script **test_zim2obsidian.py**, allowing test driven development. 

Feel free to pitch your ideas individually on this discussion forum. Although I can't promise to pick them up, someone else might.

**Comment by sojusnik** [comment]
> Maybe you'd like to pick up the spirit of the open source movement, and get involved with the code yourself?

I certainly would have done this already if coding experience was present. Also at the moment I can't afford to start learning Python due to time constraints. But I must say that with the advent of AI, to teach oneself coding with the help of ChatGPT etc. is getting more and more tempting. But first I have to clone myself to free up some time...

> Feel free to pitch your ideas individually on this discussion forum. Although I can't promise to pick them up, someone else might.

Will do it tomorrow with a detailed description and examples formatted as Zim notes to make it easier to reproduce.

As a side note, [this](https://www.youtube.com/watch?app=desktop&v=DbsAQSIKQXk) short coder-centric video does a great job in showcasing Obsidian. And [here](https://github.com/0atman/noboilerplate/blob/main/scripts/29-obsidian.md) is the supplemental material to it. You might enjoy watching it. And [this](https://www.youtube.com/watch?v=AtdAAD47aQY) one shows different approaches in organizing notes in apps like Obsidian. His channel has in general a lot of valuable information.

**Reply by peter88213** [reply]
As you can see, I've opened a discussion forum now, so this discussion could jump to a new thread in order to not go "off topic" any more.

Yesterday I installed the latest Zim version. Then I had another look at Obsidian, having converted a ~22MB Zim notebook with *zim2obsidian*. This experience brought up some new questions that I've posted in the forum now.

After all, I could not really make friends with Obsidian. May it be because of the "web app" GUI style, may it be because of bad font rendering on my PC in bright mode, may it be because I'm familiar with Zim's wiki structure that can't be represented properly with Obsidian's mere file structure (see #6). Of course, I understand that others see it differently, and I think a conversion tool with potential might well be in demand. 

As a little sporting exercise, implementing a zim2obsidian feature can be quite appealing, but as you may have seen, I have my own projects of a very different scale here on GitHub. It would not be right to raise expectations here that I cannot satisfy. No software project of any significance can survive in the long run as a one-man show. As you have seen, Zim is already on the brink because the initiator is not even capable of sifting through and evaluating the flood of feature requests. And he's a real professional doing an excellent job!

**Reply by sojusnik** [reply]
> As you can see, I've opened a discussion forum now, so this discussion could jump to a new thread in order to not go "off topic" any more.

Awesome, thanks!

> After all, I could not really make friends with Obsidian. May it be because of the "web app" GUI style, may it be because of bad font rendering on my PC in bright mode, may it be because I'm familiar with Zim's wiki structure that can't be represented properly with Obsidian's mere file structure (see https://github.com/peter88213/zim2obsidian/discussions/6).

Those were also my concerns at first, but the blazing fast development and feature richness of Obsidian outweighs the rest in this case for me.

> As a little sporting exercise, implementing a zim2obsidian feature can be quite appealing, but as you may have seen, I have my own projects of a very different scale here on GitHub. It would not be right to raise expectations here that I cannot satisfy. No software project of any significance can survive in the long run as a one-man show.

Totally understandable! Let's see where the road takes us with this endeavour.
